pywr/parameters/_parameters.pyx

```diff
@@ -306,50 +309,82 @@ cdef class TablesArrayParameter(IndexParameter):
         self.h5store = H5Store(self.h5file, None, "r")
         node = self.h5store.file.get_node(self.where, self.node)
 
-        # detect data type and read into memoryview
-        if node.dtype in (np.float32, np.float64):
-            self._values_dbl = node.read().astype(np.float64)
-            if np.min(self._values_dbl) < 0.0:
-                warnings.warn('Negative values in input file "{}" from node: {}'.format(self.h5file, self.node))
-            self._values_int = None
-            shape = self._values_dbl.shape
-        elif node.dtype in (np.int8, np.int16, np.int32):
-            self._values_dbl = None
-            self._values_int = node.read().astype(np.int32)
-            shape = self._values_int.shape
-        else:
+        if not node.dtype in (np.float32, np.float64, np.int8, np.int16, np.int32):
             raise TypeError("Unexpected dtype in array: {}".format(node.dtype))
 
+        # check the shape of the data is valid
         if self.scenario is not None:
-            if shape[1] < self.scenario.size:
+            if node.shape[1] < self.scenario.size:
                 raise RuntimeError("The length of the second dimension of the tables Node should be the same as the size of the specified Scenario.")
-            elif shape[1] > self.scenario.size:
+            elif node.shape[1] > self.scenario.size:
                 warnings.warn("The length of the second dimension of the tables Node is greater than the size of the specified Scenario. Not all data is being used!")
-        if shape[0] < len(self.model.timestepper):
+        if node.shape[0] < len(self.model.timestepper):
             raise IndexError("The length of the first dimension of the tables Node should be equal to or greater than the number of timesteps.")
 
+        # detect data type and read into memoryview
+        self._values_dbl = None
+        self._values_int = None
+        if self.scenario:
+            # if possible, only load the data required
+            scenario_indices = None
+            if self.model.scenarios.user_combinations:
+                scenario_indices = set()
+                for user_combination in self.model.scenarios.user_combinations:
+                    scenario_indices.add(user_combination[self._scenario_index])
+                scenario_indices = sorted(list(scenario_indices))
+                self._scenario_ids = np.ones(self.scenario.size, dtype=np.int32)
```

Comment:
I agree. I'm not sure what else to do though. They can't be set to None/NaN as there isn't an integer representation. They could be set to a negative number, but that's still a valid index. I guess they could be set to `node.shape[1]` so that it would at least raise an `IndexError` if it was ever used.